Source/SwiftLintBuiltInRules/Rules/Metrics/CyclomaticComplexityRule.swift

```diff
@@ -69,70 +70,109 @@ struct CyclomaticComplexityRule: ASTRule {
+            let complexity = ComplexityVisitor(
+                ignoresCaseStatements: configuration.ignoresCaseStatements
+            ).walk(tree: body, handler: \.complexity)
+
+            for parameter in configuration.params where complexity > parameter.value {
+                let reason = "Function should have complexity \(configuration.length.warning) or less; " +
+                             "currently complexity is \(complexity)"
+
+                let violation = ReasonedRuleViolation(
+                    position: violationToken.positionAfterSkippingLeadingTrivia,
+                    reason: reason,
+                    severity: parameter.severity
+                )
+                violations.append(violation)
+                return
+            }
+        }
     }
 
-    private func measureComplexity(in file: SwiftLintFile, dictionary: SourceKittenDictionary) -> Int {
-        var hasSwitchStatements = false
+    private class ComplexityVisitor: SyntaxVisitor {
+        private(set) var complexity = 0
+        let ignoresCaseStatements: Bool
 
-        let complexity = dictionary.substructure.reduce(0) { complexity, subDict in
-            guard subDict.kind != nil else {
-                return complexity
-            }
+        init(ignoresCaseStatements: Bool) {
+            self.ignoresCaseStatements = ignoresCaseStatements
+            super.init(viewMode: .sourceAccurate)
+        }
 
-            if let declarationKind = subDict.declarationKind,
-                SwiftDeclarationKind.functionKinds.contains(declarationKind) {
-                return complexity
-            }
+        override func visitPost(_ node: ForStmtSyntax) {
+            complexity += 1
+        }
 
-            guard let statementKind = subDict.statementKind else {
-                return complexity + measureComplexity(in: file, dictionary: subDict)
-            }
+        override func visitPost(_ node: IfExprSyntax) {
+            complexity += 1
+        }
 
-            if statementKind == .switch {
-                hasSwitchStatements = true
-            }
-            let score = configuration.complexityStatements.contains(statementKind) ? 1 : 0
-            return complexity +
-                score +
-                measureComplexity(in: file, dictionary: subDict)
+        override func visitPost(_ node: GuardStmtSyntax) {
+            complexity += 1
         }
 
-        if hasSwitchStatements && !configuration.ignoresCaseStatements {
-            return reduceSwitchComplexity(initialComplexity: complexity, file: file, dictionary: dictionary)
+        override func visitPost(_ node: RepeatStmtSyntax) {
+            complexity += 1
         }
 
-        return complexity
-    }
+        override func visitPost(_ node: WhileStmtSyntax) {
+            complexity += 1
+        }
 
-    // Switch complexity is reduced by `fallthrough` cases
+        override func visitPost(_ node: CatchClauseSyntax) {
```

Comment:
this is a bit weird, and maybe not intentional but this snippet:

```swift
do {} catch {}
```

contains a `case` according to SourceKitten:

```json
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.length" : 14,
  "key.offset" : 0,
  "key.substructure" : [
    {
      "key.bodylength" : 0,
      "key.bodyoffset" : 4,
      "key.kind" : "source.lang.swift.stmt.brace",
      "key.length" : 2,
      "key.offset" : 3
    },
    {
      "key.elements" : [
        {
          "key.kind" : "source.lang.swift.structure.elem.pattern",
          "key.length" : 1,
          "key.offset" : 12
        }
      ],
      "key.kind" : "source.lang.swift.stmt.case",
      "key.length" : 8,
      "key.offset" : 6
    }
  ]
}
```

since it's kind of a different branch I decided to keep counting it.

Reply:
The fact that SourceKit thinks that's a case statement is weird, but it's true that there are two branches in `do {} catch {}`: one where only `do` is invoked and another where `do` and `catch` are invoked.